bigquery/google/cloud/bigquery/magics.py

```diff
@@ -0,0 +1,251 @@
+    """Runs a query while printing status updates
+
+    Args:
+        client (google.cloud.bigquery.client.Client):
+            Client to bundle configuration needed for API requests.
+        query (str):
+            SQL query to be executed. Defaults to the standard SQL dialect.
+            Use the ``job_config`` parameter to change dialects.
+        job_config (google.cloud.bigquery.job.QueryJobConfig, optional):
+            Extra configuration options for the job.
+
+    Returns:
+        str: the ID of the query job created
+
+    Example:
+        >>> client = bigquery.Client()
+        >>> _run_query(client, "SELECT 17")
+        Executing query with job ID: bf633912-af2c-4780-b568-5d868058632b
+        Query executing: 1.66s
+        Query complete after 2.07s
+        'bf633912-af2c-4780-b568-5d868058632b'
+    """
+    start_time = time.time()
+    query_job = client.query(query, job_config=job_config)
+    print('Executing query with job ID: {}'.format(query_job.job_id))
+
+    while True:
+        print('\rQuery executing: {:0.2f}s'.format(
+            time.time() - start_time), end='')
+        try:
+            query_job.result(timeout=0.5)
+        except futures.TimeoutError:
+            continue
+        break
+    print('\nQuery complete after {:0.2f}s'.format(time.time() - start_time))
+    return query_job.job_id
+
+
+@magic_arguments.magic_arguments()
+@magic_arguments.argument(
+    'destination_var',
+    nargs='?',
+    help=('If provided, save the output to this variable in addition '
+          'to displaying it.'))
+@magic_arguments.argument(
+    '--project',
+    type=str,
+    default=None,
+    help=('Project to use for executing this query. Defaults to the '
+          'client\'s project'))
+@magic_arguments.argument(
+    '--use_legacy_sql', action='store_true', default=False,
+    help=('Sets query to use Legacy SQL instead of Standard SQL. Defaults to '
+          'Standard SQL if this argument is not used.'))
+@magic_arguments.argument(
+    '--verbose', action='store_true', default=False,
+    help=('If set, print verbose output, including the query job ID and the '
+          'amount of time for the query to finish. By default, this '
+          'information will be displayed as the query runs, but will be '
+          'cleared after the query is finished.'))
+def _cell_magic(line, query):
+    """Underlying function for bigquery cell magic
+
+        Note:
+            This function contains the underlying logic for the 'bigquery' cell
+            magic. This function is not meant to be called directly.
+
+        Args:
+            line (str): "%%bigquery" followed by arguments as required
+            query (str): SQL query to run
+
+        Returns:
+            pandas.DataFrame: the query results.
+    """
+    args = magic_arguments.parse_argstring(_cell_magic, line)
+
+    if args.project:
+        client = Client(
+            project=args.project, credentials=context.client._credentials)
```

Comment:
It sounds like this will be an issue for users relying on Application Default Credentials who do not have a default project set. These users will have to set the context client manually. Was there anything more to the error message that told you the problem was not setting the environment default project? 

I think the answer to this may be documenting that in order to have the context client set automatically, you need to have application default credentials _and_ a default project set up.

Reply:
Well, the thing that was particularly a bummer here is that I *am* passing `--project` -- so as the user, I'm like "hey I've got a project right here!"

I wonder if this means you don't want to use `context.client._credentials` if it's not yet set?

Reply:
The Client object requires a project, and we need to construct a Client to use ADC. It would be possible to set the first project used as the context client project in this situation, but I was worried that would be unclear. The table below is what the result would be for the different behaviors if I made the change. @tswast what do you think?

Behavior | Result
--- | ---
Application Default Credentials + Environment default Project | Use environment default project as default
Application Default Credentials + No environment default project + Project argument is given | Use first project used as default for session
Application Default Credentials + No environment default project + No project argument | Throw error stating that project is required

Reply:
> construct a Client to use ADC

Not necessarily. You can use `google.auth.default` directly if you need more control over how auth/project is resolved.

Reply:
+1 to Jon's comment. This is why we documented the exception thrown as a default credentials error, so that folks that need more complex auth can catch it and provide their own.

That said, maybe instead of having a Client in context, maybe we should have a Credentials object in the context so that a client can be constructed at query time?